lib/segment/src/index/hnsw_index/hnsw.rs

```diff
@@ -94,8 +94,12 @@ impl<TGraphLinks: GraphLinks> HNSWIndex<TGraphLinks> {
         } else {
             let vector_storage = vector_storage.borrow();
             let available_vectors = vector_storage.available_vector_count();
-            let full_scan_threshold = hnsw_config.full_scan_threshold.saturating_mul(BYTES_IN_KB)
-                / (vector_storage.vector_dim() * VECTOR_ELEMENT_SIZE);
+            let full_scan_threshold = if available_vectors > 0 {
+                hnsw_config.full_scan_threshold.saturating_mul(BYTES_IN_KB)
+                    / (vector_storage.available_size_in_bytes() / available_vectors)
```

Comment:
In theory, `vector_storage.available_size_in_bytes()` might be zero even if available_vectors is not. E.g. if we have empty sparse vectors or something like this. It might not happen in practice, but I would do the check anyway.

Also, please consider `checked_div`

Reply:
Good catch!

Reply:
In this case, we can probably do:

```rust
let full_scan_threshold = hnsw_config
    .full_scan_threshold
    .saturating_mul(BYTES_IN_KB)
    .checked_div(
        vector_storage
            .available_size_in_bytes()
            .checked_div(available_vectors)
            .unwrap_or_default(),
    )
    .unwrap_or(1);
```

or the other way around:

```rust
let full_scan_threshold = vector_storage
    .available_size_in_bytes()
    .checked_div(available_vectors)
    .and_then(|div| {
        hnsw_config
            .full_scan_threshold
            .saturating_mul(BYTES_IN_KB)
            .checked_div(div)
    })
    .unwrap_or(1);
```

Reply:
done, renamed `div` in snippet into `avg_vector_size`

Reply:
Better. Thanks!